src/controllers/ilp-prepare.js

```diff
@@ -0,0 +1,108 @@
+'use strict'
+
+const IlpPacket = require('ilp-packet')
+const log = require('../common').log.create('ilp-prepare')
+
+const Config = require('../services/config')
+const Accounts = require('../services/accounts')
+const RouteBuilder = require('../services/route-builder')
+const RateBackend = require('../services/rate-backend')
+const PeerProtocolController = require('../controllers/peer-protocol')
+
+const UnreachableError = require('../errors/unreachable-error')
+
+const { fulfillmentToCondition } = require('../lib/utils')
+const { codes } = require('../lib/ilp-errors')
+
+const PEER_PROTOCOL_PREFIX = 'peer.'
+
+class IlpPrepareController {
+  constructor (deps) {
+    this.config = deps(Config)
+    this.accounts = deps(Accounts)
+    this.routeBuilder = deps(RouteBuilder)
+    this.backend = deps(RateBackend)
+    this.peerProtocolController = deps(PeerProtocolController)
+  }
+
+  async handle (sourceAccount, packet) {
+    log.debug('handling ilp prepare. sourceAccount=%s', sourceAccount)
+
+    try {
+      const parsedPacket = IlpPacket.deserializeIlpPrepare(packet)
+      const { amount, executionCondition, destination } = parsedPacket
+
+      if (destination.startsWith(PEER_PROTOCOL_PREFIX)) {
+        return await this.peerProtocolController.handle(sourceAccount, parsedPacket)
+      }
+
+      const { nextHop, nextHopPacket } = await this.routeBuilder.getNextHopPacket(sourceAccount, parsedPacket)
+
+      log.debug('sending outbound ilp prepare. destination=%s amount=%s', destination, nextHopPacket.amount)
+      const result = await this.sendDataToPeer(nextHop, IlpPacket.serializeIlpPrepare(nextHopPacket))
```

Comment:
Shouldn't we modify the ILP packet in place rather than re-encoding? (That could also be a separate PR though)

Reply:
Imho, that's premature optimization. Not long ago, we were ok going with a text-based format that would have been an order of magnitude slower. It's just not worth worrying about trying to get zero-copy performance at this stage. There are much lower hanging fruit.